Data/Primitive/UnliftedArray.hs

```diff
@@ -127,90 +130,233 @@ import GHC.Base (runRW#)
 -- around unlifted primitive types. The values of the unlifted type are
 -- stored directly, eliminating a layer of indirection.
 data UnliftedArray e = UnliftedArray ArrayArray#
+#if __GLASGOW_HASKELL__ < 710
   deriving (Typeable)
+#endif
 
 -- | Mutable arrays that efficiently store types that are simple wrappers
 -- around unlifted primitive types. The values of the unlifted type are
 -- stored directly, eliminating a layer of indirection.
 data MutableUnliftedArray s e = MutableUnliftedArray (MutableArrayArray# s)
+#if __GLASGOW_HASKELL__ < 710
   deriving (Typeable)
+#endif
 
 -- | Classifies the types that are able to be stored in 'UnliftedArray' and
 -- 'MutableUnliftedArray'. These should be types that are just liftings of the
 -- unlifted pointer types, so that their internal contents can be safely coerced
 -- into an 'ArrayArray#'.
 class PrimUnlifted a where
+  {-# MINIMAL (toArrayArray# | toUnlifted#), (fromArrayArray# | fromUnlifted# ) #-}
+#if __GLASGOW_HASKELL__ >= 800
+  -- | The unlifted representation of a type.
+  --
+  -- @since TODO
+#  if __GLASGOW_HASKELL__ >= 802
+  type Unlift a :: TYPE 'UnliftedRep
+#  else
+  type Unlift a :: TYPE 'PtrRepUnlifted
+#  endif
+  type Unlift a = ArrayArray#
+#endif
+
+  -- | Convert a value to its unlifted representation.
+  --
+  -- @since TODO
```

Comment:
We should probably use `DefaultSignatures` here. That means

1. If someone defines `Unlift`, they'll also have to define `toUnlifted#` and `fromUnlifted#`, which seems reasonable.
2. We won't need the `unsafeCoerce#`.

Reply:
What signature do you have in mind? I don't see what constraint you could possibly provide.

Reply:
I just updated. I *think* I like this way better.